### DISCUSSION TITLE
How to pass an optional parameter

### DISCUSSION THREAD

**Discussion by royfrancis** [typst/typst#3076] [Q&A] [answered]
In my **template.typ**, I have this bit of code:

```
body
v(1fr)
footer
```

How do I pass the `1fr` as an optional parameter from **show.typ**? If it's empty, I want `1fr` to be the default value.

**Comment by istudyatuni** [comment]
Something like this should work

```typ
#let template(spacing: 1fr, body) = {
  body
  v(spacing)
  [Footer]
}
```

**Comment by royfrancis** [comment]
I get this error: 

```
error: expected relative length or fraction, found content
```

I guess the 1fr that I pass in seems to recognized as "content" type. Is there a way to convert content to fraction?

**Reply by istudyatuni** [reply]
Could you share your code?

**Reply by royfrancis** [reply]
In the example below, I added gap as new parameter. Without that it works.

**template.typ**

```
#let letter(
  recipient: none,
  gap: none,
  body
) = {
  set page(paper: "us-letter", margin: (top: 2cm))
  recipient
  v(gap)
  body
}
```

**show.typ**

```
#show: letter.with(
$if(recipient)$
  recipient: [$recipient$],
$endif$
$if(gap)$
  gap: [$gap$],
$endif$
)
```

letter

```
---
recipient: Mr. John Doe
gap: 2cm
format:
  letter-typst: default
---

Dear Joe,

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque tempus lorem quam, nec eleifend risus rhoncus sed. Fusce urna quam, placerat eu nisi eget, consequat tempor mauris. Nullam maximus est sit amet blandit venenatis. Curabitur vestibulum, nulla eget bibendum blandit, est nisi convallis massa, vel tincidunt orci mi dictum magna. Duis rhoncus neque vitae ipsum imperdiet pulvinar. 

Best,
Jane Smith
Regional Director
```

**Reply by istudyatuni** [reply]
So, typst uses `$if$` syntax for math, not for flow control, also read more about [scripting](https://typst.app/docs/reference/scripting). Also, when you write `#show: letter`, letter is function, which you defined above, not file, so you should write your content under this show rule. Also typst does not have frontmatter

You can configure your template in several ways

1. Pass parameters directly, so it will be your config

```typ
// template.typ
#let letter(
  recipient: none,
  gap: 1fr,
  body
) = {
  set page(paper: "us-letter", margin: (top: 2cm))
  if recipient != none {
    recipient  
  }
  v(gap)
  body
}

// show.typ
#import "template.typ": letter
#show: letter.with(
  recipient: "Mr. John Doe",
  gap: 2cm
)

Dear Joe,

#lorem(70)

Best,
Jane Smith
Regional Director
```

2. Save it as variable

```typ
// template.typ
#let letter(
  recipient: none,
  gap: 1fr,
  body
) = {
  set page(paper: "us-letter", margin: (top: 2cm))
  if recipient != none {
    recipient  
  }
  v(gap)
  body
}

// show.typ
#import "template.typ": letter

#let conf = (
  recipient: "Mr. John Doe",
  gap: 2cm,
  format: (
    letter-typst: "default"
  ),
)

#show: letter.with(
  recipient: conf.at("recipient", default: none),
  gap: conf.at("gap", default: none)
)

Dear Joe,

#lorem(70)

Best,
Jane Smith
Regional Director
```

3. You could try to load config from yaml file, but I don't know, how convert string with length, like "2сm", to typst length. Replace `#let conf` with something like:

```typ
#let conf = yaml("conf.yaml")
```

**Reply by royfrancis** [reply]
Thanks! I am using typst through quarto which is why everything looks so chaotic. I should've probably mentioned that before. Apologies! Anyhow, there must be some sort of type conversion in typst. I couldn't find anything in their [type documentation](https://typst.app/docs/reference/foundations/type/). Surely, it must be possible to convert a string to a number or concatenate a number and string to create a fraction type. `gap = 1 + "fr"`

**Reply by istudyatuni** [reply]
I got an answer in discord, that's easier with eval:

```typ
#let gap = eval("1cm")
```

**Reply by royfrancis** [reply]
How would you use that in your example? I have tested out your code on the typst website

```typst
//template

#let letter(
  recipient: none,
  gap: none,
  body
) = {
  set page(paper: "us-letter", margin: (top: 2cm))
  if recipient != none {
    recipient  
  }
  if gap != none {
    v(gap)
  }
  body
}

// show.typ

#let conf = (
  recipient: "Mr. John Doe",
  gap: 1cm,
  format: (
    letter-typst: "default"
  ),
)

#show: letter.with(
  recipient: conf.at("recipient", default: none),
  gap: conf.at("gap", default: none)
)

// body

Dear Joe,

#lorem(70)

Best,
Jane Smith
Regional Director
```

**Reply by istudyatuni** [reply, marked as answer]
Something like

```typ
#let conf = yaml("conf.yaml")
#show: letter.with(
  recipient: conf.at("recipient", default: none),
  gap: eval(conf.at("gap", default: "1cm"))
)
```
